kubeadm add support for structured ExtraArgs

neolit123 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

Add a new type Arg that holds a dedicated Name and Value.
Instead of using map[string]string for ExtraArgs in the
API use []Arg.

Adapt the kubeadm code base to the []Arg change.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref https://github.com/kubernetes/kubeadm/issues/1601

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

neolit123 committed:
kubeadm: add structured ExtraArgs in v1beta4

Add a new type Arg that holds a dedicated Name and Value.
Instead of using map[string]string for ExtraArgs in the
API use []Arg.

Adapt v1beta3 conversion to convert to/from the
legacy map[string]string.

neolit123 committed:
kubeadm: adapt the validation and utils for structured ExtraArgs

Use []kubeadm.Arg instead of map[string]string when
validating ExtraArgs in the API.

Add new GetArgValue() and SetArgValue() utilities
and tests in apis/kubeadm.

Add new utils for constucting commands from and to
a []kubeadm.Arg slice.

neolit123 committed:
kubeadm: update the rest of the code base for structured ExtraArgs

In all places map[string]string compounds were used directly
for extra args. Modify said locations to use []Arg
and the new utilities Get/SetArgValue(), ArgumentsTo/FromCommand().

neolit123 committed:
kubeadm: add TODO for NewMapStringString() and CLI extra-args

neolit123 commented:
/triage accepted
/priority important-soon

neolit123 commented on cmd/kubeadm/app/cmd/options/generic.go:
```diff
@@ -54,6 +54,9 @@ func AddIgnorePreflightErrorsFlag(fs *pflag.FlagSet, ignorePreflightErrors *[]st
 
 // AddControlPlanExtraArgsFlags adds the ExtraArgs flags for control plane components
 func AddControlPlanExtraArgsFlags(fs *pflag.FlagSet, apiServerExtraArgs, controllerManagerExtraArgs, schedulerExtraArgs *map[string]string) {
+	// TODO: https://github.com/kubernetes/kubeadm/issues/1601
+	// Either deprecate these flags or handle duplicate keys.
+	// Currently the map[string]string returned by NewMapStringString() doesn't allow this.
```
not sure what to do with this.
i think i'm +1 to deprecate these. they are a bit too complex for CLI flags - e.g. `--extra-args=foo1=bar1,foo2=bar2`

`--feature-gates` falls into the same bucket, but we actually lack config for it during upgrade, IIRC.
alternatively they need to be refactored to support the new []Arg type and not use map[string]string as the CLI flag backing type.

neolit123 replied:
i think i will just leave this TODO and keep the issue https://github.com/kubernetes/kubeadm/issues/1601 open for the time being.

pacoxu commented on cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go:
```diff
@@ -39,19 +39,55 @@ func Convert_v1beta3_InitConfiguration_To_kubeadm_InitConfiguration(in *InitConf
+		args[arg.Name] = arg.Value
+	}
+	return args
+}
```
Another contributor once asked if we need to add some tests for the `conversion.go`. 

For here, `convertFromArgs` may de-dup some keys, which is expected.

neolit123 replied:
i can add some quick unit tests. yes, converting to map drops duplicates.

chendave commented:
/cc

neolit123 commented:
/approve cancel
this is ready for review

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -663,3 +677,16 @@ func ValidateResetConfiguration(c *kubeadm.ResetConfiguration) field.ErrorList {
 	return allErrs
 }
+
+// ValidateExtraArgs validates a set of arguments and and collects all encountered errors
```
```suggestion
// ValidateExtraArgs validates a set of arguments and collects all encountered errors
```

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -142,11 +142,10 @@ type ClusterConfiguration struct {
-	// TODO: This is temporary and ideally we would like to switch all components to
-	// use ComponentConfig + ConfigMaps.
-	ExtraArgs map[string]string
+	// An argument name in this list is the flag name as it appears on the
+	// command line except without leading dash(es). Extra arguments will override existing
+	// default arguments. Duplicate extra arguments are allowed.
+	ExtraArgs []Arg
```
unclear if extra args will always remain as a config option.
removing the TODO to avoid confusion.

chendave commented:
thanks @neolit123 !

This is a big change, will take a look ASAP.

chendave commented on cmd/kubeadm/app/phases/kubelet/flags.go:
```diff
@@ -123,6 +122,6 @@ func writeKubeletFlagBytesToDisk(b []byte, kubeletDir string) error {
 
 // buildKubeletArgMap takes a kubeletFlagsOpts object and builds based on that a string-string map with flags
```
```suggestion
// buildKubeletArgMap takes a kubeletFlagsOpts object and builds based on that a slice of `Arg`
```

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -263,9 +262,10 @@ type LocalEtcd struct {
+	// An argument name in this list is the flag name as it appears on the
+	// command line except without leading dash(es). Extra arguments will override existing
+	// default arguments. Duplicate extra arguments are allowed.
+	ExtraArgs []Arg
```
Do we have any testcase to test that config file now allows the arg with duplicated name? I haven't seen it.

neolit123 replied:
i don't think we need to test it on the config structs directly, but we do test TestArgumentsToCommand, which is the important function that produces a command with --flags.

chendave reviewed (commented):
My first pass of review, mostly LGTM.

chendave commented:
pls don't forget to update [docs](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go#L30) :)

neolit123 commented on cmd/kubeadm/app/util/arguments.go:
```diff
@@ -24,41 +24,54 @@ import (
-		argsMap[k] = v
+	// Copy the base arguments into a new slice.
+	args := make([]kubeadmapi.Arg, len(base))
+	copy(args, base)
```
i am trying to understand how that would be different. can you please provide some examples? the idea was to allow a single --foo flag to remove all instances of --foo in the base, so that extra args do a true override.

chendave replied:
this saves us some time of removing the element from the `base`.

```glang
args = append(args[:j], args[j+1:]...)
```

we still need to append after searching and comparison.

neolit123 replied:
as mentioned here:
https://github.com/kubernetes/kubernetes/pull/119156/files#r1264666750
you can send a follow up with the optimizations if you'd like.

SataQiu replied:
+1 for using a set to track overrides flags, and just copy the flags that are not overwritten.

neolit123 commented on cmd/kubeadm/app/util/arguments.go:
```diff
@@ -24,41 +24,54 @@ import (
 // with optional override arguments. In the return list override arguments will precede base
-// arguments
-func BuildArgumentListFromMap(baseArguments map[string]string, overrideArguments map[string]string) []string {
+// arguments. If an argument is present in the overrides, it will cause
```
no, it's "present"

neolit123 commented on cmd/kubeadm/app/util/arguments.go:
```diff
@@ -24,41 +24,54 @@ import (
+				// Remove this existing argument and search for another argument
+				// with the same name in base.
+				args = append(args[:j], args[j+1:]...)
+				goto repeat
```
i didn't think about optimization much. maybe we can send followup PR.

chendave replied:
sure, we can leave this as a follow up.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/argument.go:
```diff
@@ -0,0 +1,67 @@
+// one if missing. The function works backwards in the argument list.
+// nArgs holds how many existing arguments with this name should be set.
+// If nArgs is negative value all of them will be updated.
+// If nArgs is zero none will be updated.
```
i just added the function and was thinking, we might need it for A, B, C. but maybe we can just simplify it.

neolit123 replied:
i guess it can be `if nArgs < 1 { /* update all */ }`

SataQiu commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -64,6 +64,9 @@ func ValidateClusterConfiguration(c *kubeadm.ClusterConfiguration) field.ErrorLi
 	allErrs = append(allErrs, ValidateDNS(&c.DNS, field.NewPath("dns"))...)
 	allErrs = append(allErrs, ValidateNetworking(c, field.NewPath("networking"))...)
 	allErrs = append(allErrs, ValidateAPIServer(&c.APIServer, field.NewPath("apiServer"))...)
+	allErrs = append(allErrs, ValidateControlPlaneComponent(&c.APIServer.ControlPlaneComponent, field.NewPath("apiServer"))...)
```
How about merging this into `ValidateAPIServer` ?

neolit123 replied:
yes, we could use the dedicated functions.

SataQiu commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -64,6 +64,9 @@ func ValidateClusterConfiguration(c *kubeadm.ClusterConfiguration) field.ErrorLi
 	allErrs = append(allErrs, ValidateAPIServer(&c.APIServer, field.NewPath("apiServer"))...)
+	allErrs = append(allErrs, ValidateControlPlaneComponent(&c.APIServer.ControlPlaneComponent, field.NewPath("apiServer"))...)
+	allErrs = append(allErrs, ValidateControlPlaneComponent(&c.ControllerManager, field.NewPath("controllerManager"))...)
+	allErrs = append(allErrs, ValidateControlPlaneComponent(&c.Scheduler, field.NewPath("sachduler"))...)
```
How about creating new functions `ValidateControllerManager` and `ValidateScheduler`?

SataQiu reviewed (commented):
Overall SGTM
add some comments for nits.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/argument.go:
```diff
@@ -0,0 +1,67 @@
+// If the argument does not exist an empty string and -1 are returned.
+// startIdx defines where the iteration starts. If startIdx is a negative
+// value or larger than the size of the argument slice the iteration
+// will start from the last element.
```
actually, the reason for this is that in a flag list (as per Cobra) the last flag takes precedence if that flag is not supported multiple times.

e.g.
--a=foo
--a=bar

if mutliple instances of --a are not allowed, then Cobra will assign bar to --a.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/argument_test.go:
```diff
@@ -0,0 +1,131 @@
+			testName:      "argument exists with non-empty value (offset index)",
+			args:          []Arg{{Name: "a", Value: "a1"}, {Name: "b", Value: "b1"}, {Name: "c", Value: "c1"}},
+			name:          "b",
+			expectedValue: "b1",
```
added better test here.

neolit123 commented:
addressed most of the comments. thank for the reviews thus far. PTAL again if you have time.

@chendave i left some of the comments about iterations / optimizations un-adressed.
i can either fix them later in this PR or you can send a new one if you'd like.

chendave commented:
> @chendave i left some of the comments about iterations / optimizations un-adressed.

sure, np, I will take another look in the following few days.

chendave commented:
/assign
(so that I can find this one easily)

chendave commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -302,62 +305,63 @@ func getControllerManagerCommand(cfg *kubeadmapi.ClusterConfiguration) []string
 	if cfg.ClusterName != "" {
-		defaultArguments["cluster-name"] = cfg.ClusterName
+		defaultArguments = kubeadmapi.SetArgValues(defaultArguments, "cluster-name", cfg.ClusterName, 1)
+
```
remove this blank line

chendave commented on cmd/kubeadm/app/phases/etcd/local.go:
```diff
@@ -239,44 +239,44 @@ func getEtcdCommand(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.A
+	defaultArguments := []kubeadmapi.Arg{
+		{Name: "name", Value: nodeName},
+		// TODO, start using --initial-corrupt-check once the graduated flag is available,
+		// https,//github.com/kubernetes/kubeadm/issues/2676
```
```suggestion
		// https://github.com/kubernetes/kubeadm/issues/2676
```

chendave commented on cmd/kubeadm/app/phases/kubelet/flags.go:
```diff
@@ -66,31 +66,30 @@ func WriteKubeletDynamicEnvFile(cfg *kubeadmapi.ClusterConfiguration, nodeReg *k
-func buildKubeletArgMapCommon(opts kubeletFlagsOpts) map[string]string {
-	kubeletFlags := map[string]string{}
-	kubeletFlags["container-runtime-endpoint"] = opts.nodeRegOpts.CRISocket
+func buildKubeletArgMapCommon(opts kubeletFlagsOpts) []kubeadmapi.Arg {
```
suggest to rename this method, as this is not building a map anymore.

chendave commented on cmd/kubeadm/app/phases/kubelet/flags.go:
```diff
@@ -121,8 +120,8 @@ func writeKubeletFlagBytesToDisk(b []byte, kubeletDir string) error {
+// buildKubeletArgMap takes a kubeletFlagsOpts object and builds based on that a slice of arguments
 // that should be given to the local kubelet daemon.
-func buildKubeletArgMap(opts kubeletFlagsOpts) map[string]string {
+func buildKubeletArgMap(opts kubeletFlagsOpts) []kubeadmapi.Arg {
```
rename this method as well.

neolit123 replied:
makes sense!

chendave commented on cmd/kubeadm/app/phases/etcd/local.go:
```diff
@@ -239,44 +239,44 @@ func getEtcdCommand(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.A
-		"experimental-watch-progress-notify-interval": "5s",
+	defaultArguments := []kubeadmapi.Arg{
+		{Name: "name", Value: nodeName},
+		// TODO, start using --initial-corrupt-check once the graduated flag is available,
```
```suggestion
		// TODO: start using --initial-corrupt-check once the graduated flag is available,
```

chendave reviewed (commented):
some more nits, otherwise, this change LGTM.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion_test.go:
```diff
@@ -0,0 +1,95 @@
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			actual := convertFromArgs(tc.args)
+			if !reflect.DeepEqual(convertToArgs(tc.expectedArgs), convertToArgs(actual)) {
```
i wasn't sure if deepequal() will have troubles with key ordering, as it not guaranteed in golang.
but it seems it always works.

```
	m1 := map[string]string{"a": "a1", "c": "c1", "b": "b1"}
	m2 := map[string]string{"a": "a1", "b": "b1", "c": "c1"}
	reflect.DeepEqual(m1, m2) // always true
```

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion_test.go:
```diff
@@ -0,0 +1,95 @@
+				{Name: "c", Value: "d1"},
+				{Name: "c", Value: "d2"},
+			},
+			expectedArgs: map[string]string{"a": "b", "c": "d2"},
```
readers can just click convertFromArgs() to see its source code.
it's pretty straight forward, the same key will override previous key.

it's better to add comments over the two convert* functions.

neolit123 commented on cmd/kubeadm/app/util/arguments.go:
```diff
@@ -24,41 +24,54 @@ import (
+// arguments. If an argument is present in the overrides, it will cause
+// all instances of the same argument in the base list to be discarded, leaving
+// only the instances of this argument in the overrides to be applied.
+func ArgumentsToCommand(base []kubeadmapi.Arg, overrides []kubeadmapi.Arg) []string {
```
well, this is not necessarily a strict rule out there.

my preference would be have all utils added in one commit:
https://github.com/kubernetes/kubernetes/pull/119156/commits/03b253e931929dd4f47b9ae3510b0aa573cb9049

and then update all code / usage in a separate one.
https://github.com/kubernetes/kubernetes/pull/119156/commits/01520623e6e47ce3e9ba3c229d5b384cd2ca5a48

chendave replied:
Okay, LGTM.

neolit123 commented on cmd/kubeadm/app/util/arguments_test.go:
```diff
@@ -73,13 +106,13 @@ func TestBuildArgumentListFromMap(t *testing.T) {
-				"something-that-allows-empty-string": "",
+			overrides: []kubeadmapi.Arg{
+				{Name: "admission-control", Value: "NamespaceLifecycle,LimitRanger"},
+				{Name: "something-that-allows-empty-string", Value: ""},
```
IIRC, `--foo=` is a valid empty string flag value and there are some usages in k8s for it.

chendave commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -663,3 +684,16 @@ func ValidateResetConfiguration(c *kubeadm.ResetConfiguration) field.ErrorList {
+
+	for idx, arg := range args {
+		if len(arg.Name) == 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath, fmt.Sprintf("index %d", idx), "argument has no name"))
```
SGTM

chendave reviewed (commented):
/lgtm

thanks for addressing my comments!

chendave commented:
@neolit123 need rebase to fix merge conflict.

chendave commented:
/retest
/lgtm

neolit123 commented:
/assign @SataQiu 
PTAL for another review pass / approve

SataQiu reviewed (commented):
Great feature!

/approve

neolit123 commented:
/retest

neolit123 commented:
added TODO about arguments.go optimization inhttps://github.com/kubernetes/kubeadm/issues/1601

neolit123 commented:
> storage test suite panic. /retest

https://github.com/kubernetes/kubernetes/issues/120080

neolit123 commented:
/retest

more storage flakes
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/119156/pull-kubernetes-e2e-gce/1693541286981668864/
